src/Scope/TreeTraverser.php

```diff
@@ -0,0 +1,193 @@
+            }
+
+            $childScope = $this->getScopeInChild($node, $childName, $scope);
+
+            if (\is_array($child)) {
+                foreach ($child as $actualChild) {
+                    $this->traverseRecursive($actualChild, $visitor, $childScope);
+                }
+            } else {
+                $this->traverseRecursive($child, $visitor, $childScope);
+            }
+        }
+
+        $this->modifyScopeAfterNode($node, $scope);
+    }
+
+    /**
+     * E.g. in function body, gets the scope consisting of parameters and used names.
+     *
+     * @return Scope
+     *   The new scope, or the same scope instance if the child does not has its own scope.
+     */
+    private function getScopeInChild(Node $node, string $childName, Scope $scope): Scope
+    {
+        if ($node instanceof FunctionLike
+            && $childName === 'compoundStatementOrSemicolon'
+            && $node->compoundStatementOrSemicolon instanceof Node\Statement\CompoundStatementNode
+        ) {
+            $childScope = new Scope;
+            $childScope->currentSelf = $scope->currentSelf;
+            $childScope->resolvedNameCache = $scope->resolvedNameCache;
+            $isStatic = $node instanceof Node\MethodDeclaration ? $node->isStatic() : !empty($node->staticModifier);
+            if (!$isStatic && isset($scope->variables['this'])) {
+                $childScope->variables['this'] = $scope->variables['this'];
+            }
+
+            if ($node->parameters !== null) {
+                foreach ($node->parameters->getElements() as $param) {
+                    $childScope->variables[$param->getName()] = new Variable(
+                        // Pass the child scope when getting parameters - the outer scope cannot affect
+                        // any parameters of the function declaration.
+                        $this->definitionResolver->getTypeFromNode($param, $childScope),
+                        $param
+                    );
+                }
+            }
+
+            if ($node instanceof Node\Expression\AnonymousFunctionCreationExpression
+                && $node->anonymousFunctionUseClause !== null
+                && $node->anonymousFunctionUseClause->useVariableNameList !== null) {
+                foreach ($node->anonymousFunctionUseClause->useVariableNameList->getElements() as $use) {
+                    $name = $use->getName();
+                    // Used variable in an anonymous function. Same as parent type, Mixed if not defined in parent.
+                    $childScope->variables[$name] = new Variable(
+                        isset($scope->variables[$name]) ? $scope->variables[$name]->type : new Types\Mixed_,
+                        $use
+                    );
+                }
+            }
+
+            return $childScope;
+        }
+
+        if ($node instanceof ClassLike
+            && (in_array($childName, ['classMembers', 'interfaceMembers','traitMembers'], true))
```

Comment:
Do we need to `strictly` compare here?

Reply:
Not really - I thought it would be good form, just like using `===` instead of `==` - do you think there's a reason to change it?

Reply:
Somehow my benchmarks are showing different results:
https://gist.github.com/jens1o/621c1307ee0b9d618839211688d46dba

```php
PS C:\nginx\html\projects\php7.2-playground> php .\in_array_strict_benchmark.php
Took 0.42s with strict compare
Took 0.35s without strict compare
Took 5.47s with strict compare
Took 0.43s without strict compare
Took 0.54s with strict compare
Took 0.47s without strict compare
```

Reply:
Your benchmark shows searching for an empty string which does not is the case here?

Reply:
@nikic is it expexcted that strictly searching for any empty string via in_array  is slower then when doing it non-strict?

Reply:
> Your benchmark shows searching for an empty string which does not is the case here?

I didn't knew what I could use there.

Reply:
@staabm Well, that *specific* benchmark is rather useless, as the second array contains 0 as first element, which is (non-strict) equal to the empty string, so what you see there is the difference between matching on the first element in the array vs scanning through all 10000 elements and not finding anything.

Reply:
That said, the in_array strict path is slightly less optimized than the non-strict one. Though if you want to micro-optimize this, the way to do it is to use `\in_array` (with strict), which will transform to an HT lookup. Or even better, just directly implement it as an HT lookup so you're not dependent on optimization behavior.

Reply:
Ah, I didn't actually mean performance-wise but style-wise.

I think it's better because the code does not perform convoluted type casts under the hood. It's not that the code would be more incorrect that way, just that it *feels* that way - when I'm searching for a bug, I don't have to wonder if this comparison here does something funny.

Reply:
... although I would have expected the strict version also to be faster.

Reply:
> Well, that specific benchmark is rather useless, as the second array contains 0 as first element, which is (non-strict) equal to the empty string, so what you see there is the difference between matching on the first element in the array vs scanning through all 10000 elements and not finding anything.

Good point!

Reply:
I agree that strict comparison is always easier to reason about.